autopilot/contractor.go

```diff
@@ -633,7 +646,34 @@ func (c *contractor) renewFundingEstimate(ctx context.Context, cfg api.Autopilot
 	return cappedEstimatedCost, nil
 }
 
-func (c *contractor) candidateHosts(ctx context.Context, cfg api.AutopilotConfig, used map[types.PublicKey]bool, wanted int) ([]hostdb.Host, error) {
+func (c *contractor) managedFindMinAllowedHostScores(ctx context.Context, cfg api.AutopilotConfig, storedData map[types.PublicKey]uint64, redundancy float64) (float64, error) {
+	// Pull a new set of hosts from the hostdb that could be used as a new set
+	// to match the allowance. The lowest scoring host of these new hosts will
+	// be used as a baseline for determining whether our existing contracts are
+	// worthwhile.
+	numContracts := cfg.Contracts.Amount
+	buffer := 50
+	hosts, err := c.candidateHosts(ctx, cfg, make(map[types.PublicKey]struct{}), storedData, int(numContracts)+int(buffer), 1) // 1 to avoid 0 score hosts
```

Comment:
`candidateHosts` is really expensive and we call it twice now (potentially) - kind of unfortunate that every iteration fetches all hosts from the database twice... can we avoid it by keeping track of host scores? So essentially every host we ever encounter contributes to the lowest host score, and we don't calculate `managedFindMinAllowedHostScores` explicitly?

Reply:
I changed it to fetch the hosts only once. That's a bit simpler. With your new pruning code things should become a lot lightweight as well.

Reply:
agreed, I just think we shouldn't necessarily rely on pruning and ensure we don't do this twice - thanks for updating it